src/stream/src/executor/over_window/over_partition.rs

```diff
@@ -0,0 +1,1256 @@
+        }
+        // len >= 2
+        let cache_inner = self.range_cache.inner();
+        let sentinels = [
+            // sentinels only appear at the beginning and/or the end
+            cache_inner.first_key_value().unwrap().0.is_sentinel(),
+            cache_inner.last_key_value().unwrap().0.is_sentinel(),
+        ];
+        len - sentinels.into_iter().filter(|x| *x).count()
+    }
+
+    fn cache_real_first_key(&self) -> Option<&StateKey> {
+        self.range_cache
+            .inner()
+            .iter()
+            .find(|(k, _)| k.is_normal())
+            .map(|(k, _)| k.as_normal_expect())
+    }
+
+    fn cache_real_last_key(&self) -> Option<&StateKey> {
+        self.range_cache
+            .inner()
+            .iter()
+            .rev()
+            .find(|(k, _)| k.is_normal())
+            .map(|(k, _)| k.as_normal_expect())
+    }
+
+    fn cache_left_is_sentinel(&self) -> bool {
+        self.range_cache
+            .first_key_value()
+            .map(|(k, _)| k.is_sentinel())
+            .unwrap_or(false)
+    }
+
+    fn cache_right_is_sentinel(&self) -> bool {
+        self.range_cache
+            .last_key_value()
+            .map(|(k, _)| k.is_sentinel())
+            .unwrap_or(false)
+    }
+
+    /// Write a change record to state table and cache.
+    /// This function must be called after finding affected ranges, which means the change records
+    /// should never exceed the cached range.
+    pub fn write_record(
+        &mut self,
+        table: &mut StateTable<S>,
+        key: StateKey,
+        record: Record<OwnedRow>,
+    ) {
+        table.write_record(record.as_ref());
+        match record {
+            Record::Insert { new_row } | Record::Update { new_row, .. } => {
+                self.range_cache.insert(CacheKey::from(key), new_row);
+            }
+            Record::Delete { .. } => {
+                self.range_cache.remove(&CacheKey::from(key));
+
+                if self.cache_real_len() == 0 && self.range_cache.len() == 1 {
+                    // only one sentinel remains, should insert the other
+                    self.range_cache
+                        .insert(CacheKey::Smallest, OwnedRow::empty());
+                    self.range_cache
+                        .insert(CacheKey::Largest, OwnedRow::empty());
+                }
+            }
+        }
+    }
+
+    /// Find all ranges in the partition that are affected by the given delta.
+    /// The returned ranges are guaranteed to be sorted and non-overlapping. All keys in the ranges
+    /// are guaranteed to be cached.
+    pub async fn find_affected_ranges<'s, 'cache>(
+        &'s mut self,
+        table: &'_ StateTable<S>,
+        delta: &'cache PartitionDelta,
+    ) -> StreamExecutorResult<(
+        DeltaBTreeMap<'cache, CacheKey, OwnedRow>,
+        Vec<(
+            &'cache CacheKey,
+            &'cache CacheKey,
+            &'cache CacheKey,
+            &'cache CacheKey,
+        )>,
```

Comment:
Return keys with lifetime `'cache` which is corresponding to parameter `delta` and `self` (`'s: 'cache`) to give the caller confidence that the key is guaranteed to exist in the `DeltaBTreeMap`.